### ISSUE TITLE
FR: add support to disable touchscreen on Pocketbook inkpad 3

### ISSUE THREAD

**Issue by lilygthb** [koreader/koreader#9695]
**Problem**
I often read holding the reader with my thumb resting on the touchscreen. This activates the touchscreen, starting a highlight or dictionary definition. I could disable highlighting, but I would need to enable it again when I need it.

I have to do this frequently (I usually read a fiction (no highlights) and a non fiction (to be highlighted) book at the same time) and going into the menu and disabling/enabling highlighting every time I switch is annoying.

---

**Solution**
- The stock Pocketbook inkpad  reader can disable/enable touchscreen input via a double click of one of the physical buttons, It would be useful to be able to do that in Koreader as well (page turn would be done via buttons).

---

**Alternative options**
- create a Highlight on/off toggle to be assigned to a gesture (at the moment I only see a Highlight cycle option). This way I could create a specific gesture to switch between highlight/no highlight books.
- create a disable/enable touchscreen toggle to be assigned to a gesture.

**Note**
This feature could be useful also for _people with reduced mobility_ and/or difficulty holding the device, since would open more way to hold the reader without accidentally turning pages or highlighting words.

**Comment by NiLuJe** [comment]
The framework is (mostly) there between Dispatcher & the inhibitInput stuff in Input, it just needs some massaging to fit this sort of use-cases.

**Comment by pazos** [comment]
There's also something very specific implemented for very specific android devices: on devices with a "Camera" button pressing it will toggle touschreen events.

This happens on android before events are relayed to base and looks like / is a quick hack, but maybe we can do something like that with the building blocks @NiLuJe talked about and make it platform agnostic:

1. A new event that toggles inhibitInput.
2. The event is never emitted by core/plugins
3. Let users with specific needs to remap that new event in one of the spare buttons  of their devices.

I think that will fix the issue for OP (with the drawback of loosing the menu button). Anyways, devices without buttons are not suitable for this kind of FR.

**Comment by NiLuJe** [comment]
The issue with that approach is that we don't really have spare buttons on standard eInk devices ;o).

I was thinking of just doing both page-turn buttons, but that probably won't work because we react on keydown, not keyup, and I'm not even sure these events would happen across the same input frame.

I have some notes for stuff to try, but won't have time to look into it any time soon.

**Comment by Frenzie** [comment]
I think it makes sense well enough on my Kobo, I just can't be bothered to write something that's actually usable for a larger audience atm.

**Comment by lilygthb** [comment]
kindle standard software has an option to turn off touchscreen inputs as well. At least a couple of years ago it was triggered via a menu item, and to disable it you had to put the kindle to sleep and wake it again.

Having a physical button to activate inhibitinput is better but the kindle way is an option.

Another option could be inhibit all input from touchscreen, apart form a corner, which, if long pressed, could restore touchscreen inputs.

please note I cannot code, I'm just commenting from the UX point of view, I might propose something that doesn't make sense for a programmer.

**Comment by sebastien-maillard** [comment]
Hey @lilygthb , I know your issue has been resolved for a while, but I'm gonna try my luck!
Since your request has been implemented, did you manage to map a physical button to disable the touchscreen in koreader? I'd like to replicate the default mapping of the stock reader, i.e. double press the menu button, but I'm a bit lost.
I'm guessing I should create a new [key mapping](wiki/Keymapping) but I'm not sure where to find the values for the physical buttons of the InkPad 3 and their events (simple/double/long press).
Also, what should be the associated action, is it `IgnoreTouchInput` following the new event that was created?

**Comment by sebastien-maillard** [comment]
Thanks for the feedback @pazos ! Indeed, I suspected that the double press would be tricky to map for that very reason. I'll just assign the action to a simple button press then.

That's assuming I find the integer values associated to each button. So far, I haven't found any resource documenting that point.

**Comment by pazos** [comment]
> That's assuming I find the integer values associated to each button. So far, I haven't found any resource documenting that point.

That's because it is fixed for each platform. See pocketbook buttons in https://github.com/koreader/koreader/blob/master/frontend/device/pocketbook/device.lua#L191-L200 and their numbers starting at https://github.com/koreader/koreader-base/blob/master/ffi/inkview_h.lua#L815

Documentation is always welcome. Feel free to update https://github.com/koreader/koreader/wiki/Keymapping to make it less kobo specific or just add an example of how you setup your keys on your pocketbook.

**Comment by sebastien-maillard** [comment]
> That's because it is fixed for each platform. See pocketbook buttons in https://github.com/koreader/koreader/blob/master/frontend/device/pocketbook/device.lua#L191-L200 and their numbers starting at https://github.com/koreader/koreader-base/blob/master/ffi/inkview_h.lua#L815

Amazing, thanks! I did stumble upon `pocketbook/device.lua` but I couldn't find where the `C.KEY_*` constants were defined in the `koreader` project. I saw [this thread](issues/12703) mentioning `inkview_h.lua`, but that's how far I got.  I didn't realize they came from the `koreader-base` project, thank you so much for putting me on the right track!

I specifically mentioned that thread because [it seems](issues/12703#issuecomment-2454538535) the latest version of the PocketBook SDK introduced some breaking changes with the integer values associated to some of the keys, so I'll make sure to take that into consideration.

> Documentation is always welcome. Feel free to update https://github.com/koreader/koreader/wiki/Keymapping to make it less kobo specific or just add an example of how you setup your keys on your pocketbook.

Definitely! I'm a total newbie when it comes to customizing KOReader and I definitely struggled connecting the dots when reading the doc. I'll make sure to improve the documentation as soon as I figured out the mapping 🚀

**Comment by sebastien-maillard** [comment]
Since I'm getting slightly more familiar with the KOReader codebase and some of the concepts related to customization, I took another look at [the commit](https://github.com/koreader/koreader/commit/1ec6fb5fcfe074587b664c0a386680d9108283e8) associated with this request.
This made me realize the solution implemented is actually not based on some key binding, instead it introduced a toggle (_Toggle touch input_) that can be mapped to a specific gesture. I successfully managed to bind the toggle and the feature (disable/enable touch input) does work as expected, nice!

I've been playing around with key mapping and, while I could successfully modify the bindings, I still have a few questions.

1. My intent was to use the _Menu_ physical button to toggle touch input, similarly to the solution suggested in this request. I was able to bind a different event with the `Menu` key (for instance the `Home` event), which confirms I'm using the correct key, e.g.:
```
return {
            [23] = "Home", -- KEY_MENU
 }
```
However, since the feature was implemented as a "gesture" toggle (at least that's how I understand it), is it possible to bind that "action" to a key? I've tried binding `"IgnoreTouchInput"` (the [event's name](https://github.com/koreader/koreader/commit/1ec6fb5fcfe074587b664c0a386680d9108283e8#diff-73231974e9f228d37b8cd3abe7abf22d9d75e4d04e378c5c6de575e0fc8396bfR85)) and `"toggle_touch_input"` (the [menu item's name](https://github.com/koreader/koreader/commit/1ec6fb5fcfe074587b664c0a386680d9108283e8#diff-73231974e9f228d37b8cd3abe7abf22d9d75e4d04e378c5c6de575e0fc8396bfR241), though this one was definitely a long shot) and both failed, i.e. pressing the _Menu_ button doesn't do anything.

2. More generally, is it possible to bind an action outside of the ones that are [already mapped](https://github.com/koreader/koreader/blob/master/frontend/device/pocketbook/device.lua#L188C1-L201C60) with the keys? I was basically able to swap existing bindings, e.g. `Menu` key triggers `Home` event, `Next` key triggers `Menu` event, etc. But I couldn't figure out how to bind an action that wasn't already mapped to a key.

I think I'm not getting something here. There's probably some kind of distinction between an event ("_Next_ button pressed") and the action itself, which depends on the context ("Go to next page" when a document is open, "Display next 15 files" when in browser mode, etc.).
If that's the case, `event_map.lua` would probably be limited to only swapping **existing** key mappings. In other words, it wouldn't be possible to customize the actions associated to physical buttons, unlike what's possible with gestures for instance.

3. Additionally, is it possible to map keys which are not mapped? In particular, I thought of binding the `Menu` event with the _Power_ key (after making sure the _Power_ button doesn't have any default action in the InkPad settings), e.g.:
```
return {
            [1] = "Menu", -- KEY_POWER
}
```
However, this doesn't seem to have any effect. I might be wrong with the integer value though, but [based on to the SDK](https://github.com/pocketbook/SDK_6.3.0/blob/dd8de463733a2bc0d53d1214b64734586fac427e/SDK-B288/usr/arm-obreey-linux-gnueabi/sysroot/usr/local/include/inkview.h#L480), it should probably be `1` (as in `0x01`).

Sorry for the long message. Plus it's starting to get off topic, especially since the feature itself works as expected. My main goal at this point is to better understand key mapping in order to improve the documentation. I'd be happy to move this discussion to a more suitable place, in case you have some suggestions!

PS: thanks @benoit-pierre for editing the links I posted, I'll be more careful from now on ;)

**Comment by sebastien-maillard** [comment]
Thanks @NiLuJe for the feedback, that was really helpful! Documenting my findings below, in case someone else wants to look into this type of customization.

> Check verbose debug logs for the proper key events (an maybe wait for the rework to be in so that potentially new events make it in).

Indeed, after [enabling the logs](https://github.com/koreader/koreader/issues/2530#issuecomment-500866288), pressing the buttons trigger `PB event` and `key event` logs. Unfortunately, pressing the _Power_ button doesn't seem to trigger any log.

> You can then maybe try to cobble something up with both `event_map` and `event_map_adapter` + `UIManager.event_handlers`, to automagically make a Key event send an actual (... sort of) Event (e.g., see how Kobo uses it to implement custom key events).

I tried uploading a `event_map_adapter.lua` file to the settings folder, but it doesn't seem to be taken into account. So I'm guessing I'd have to directly patch the code of [`device.lua`](https://github.com/koreader/koreader/blob/master/frontend/device/pocketbook/device.lua) to include an `event_map_adapter` code block after the [`event_map`](https://github.com/koreader/koreader/blob/fd3727274b44a7ba6531aa778189e0ab3a6e05fa/frontend/device/pocketbook/device.lua#L191C9-L191C18) block, similarly to [Kobo's](https://github.com/koreader/koreader/blob/fd3727274b44a7ba6531aa778189e0ab3a6e05fa/frontend/device/kobo/device.lua#L838).
Also, it seems the main purpose of the adapter is to distinguish `isEvKeyPress` and `isEvKeyRelease` events. I'm guessing for my use-case, I could simply associate a custom value to the `C.KEY_MENU` key (e.g. `"IgnoreTouchInput"`), skip the adapter and go straight to the `UIManager.event_handlers` part.

At this point, I'm guessing the next step is to implement a `PocketBook:setEventHandlers(uimgr)` function in `device.lua`, similarly to [Kobo's](https://github.com/koreader/koreader/blob/fd3727274b44a7ba6531aa778189e0ab3a6e05fa/frontend/device/kobo/device.lua#L1625). And handle the `IgnoreTouchInput` event to replicate the behavior of [`InputContainer:setIgnoreTouchInput(state)`](https://github.com/koreader/koreader/blob/fd3727274b44a7ba6531aa778189e0ab3a6e05fa/frontend/ui/widget/container/inputcontainer.lua#L312), e.g.:
```
function PocketBook:setEventHandlers(uimgr)
    UIManager.event_handlers.IgnoreTouchInput = function()
        -- Implement enable/disable touch input feature here
    end
end
```

TL;DR It seems patching the code is the way to go. I was hoping there would be a "no-code" way to change the actions associated to buttons, but that doesn't seem to be possible indeed.

> Documentation is always welcome. Feel free to update https://github.com/koreader/koreader/wiki/Keymapping to make it less kobo specific or just add an example of how you setup your keys on your pocketbook.

I [updated](https://github.com/koreader/koreader/wiki/Keymapping/8a62e4c2e13770dab09f1418a0266d0ecfc25ce5) the documentation accordingly!

**Comment by lilygthb** [comment]
> Hey @lilygthb , I know your issue has been resolved for a while, but I'm gonna try my luck!
> 
> Since your request has been implemented, did you manage to map a physical button to disable the touchscreen in koreader? I'd like to replicate the default mapping of the stock reader, i.e. double press the menu button, but I'm a bit lost.
> 
> I'm guessing I should create a new [key mapping](https://github.com/koreader/koreader/wiki/Keymapping) but I'm not sure where to find the values for the physical buttons of the InkPad 3 and their events (simple/double/long press).
> 
> Also, what should be the associated action, is it `IgnoreTouchInput` following the new event that was created?

Sorry for the late reply, I never managed to use a physical button to disable the touch, I've mapped to a gesture:

`"cog" settings -> taps and gestures -> Gesture Manager -> long press on toggle -> bottom right (toggle touch input)`

I must say that I can never thank enough the people that actually implemented my suggestion, it really changed how I can read at night

**Comment by sebastien-maillard** [comment]
> I've mapped to a gesture

Same here! Thanks for taking the time to reply :)

> I must say that I can never thank enough the people that actually implemented my suggestion

100%
